design/ceph/ceph-telemetry.md

```diff
@@ -0,0 +1,133 @@
+# Adding Rook data to Ceph telemetry
+The solution plan agreed upon with the telemetry team is for the Rook operator to add telemetry to
+the Ceph mon `config-key` database, and Ceph will read each of those items for telemetry retrieval.
+
+## Guidelines from the Ceph telemetry team
+- Users must opt in to telemetry
+- Users must also opt into the new (delta) telemetry items added between Ceph versions
+- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
+  usage of the mon database low (limited growth is still acceptable)
+
+## Metrics for Rook to collect
+Metric names will indicate a hierarchy that can be parsed to add it to Ceph telemetry collection in
+a more ordered fashion.
+
+For example `rook/version` and `rook/kubernetes/version` would be put into a structure like shown:
+```json
+"rook": {
+  "version": "vx.y.z"
+  "kubernetes": {
+    "version": "vX.Y.Z"
+  }
+}
+```
+
+### Overall metrics
+- `rook/version` - Rook version.
+- `rook/kubernetes/...`
+  - `rook/kubernetes/version` - Kubernetes version.
+    - Ceph already collects os/kernel versions.
+  - `rook/kubernetes/node/count/total` - Number of Kubernetes nodes.
+- `rook/csi/...`
+  - `rook/csi/version` - Ceph CSI version.
+  - `rook/csi/plugin-pod/...` - Info about CSI plugin pods
+    - `rook/csi/plugin-pod/count/...` - Info about how many CSI plugin pods exist and of what types
+      This gives idea of how many k8s nodes are in the k8s cluster that are allowed to use Ceph
+      storage. With different types, telemetry can show if some CSI drivers are ever disabled.
+      - `rook/csi/plugin-pod/count/max` - Max number of pods of any CSI plugin type
+      - `rook/csi/plugin-pod/count/types` - Number of CSI plugin types. This can helpfully show if a
+        newer version of Rook supports a new plugin pod type than the Ceph telemetry version is
+        aware of.
+      - `rook/csi/plugin-pod/count/rbd`
+      - `rook/csi/plugin-pod/count/cephfs`
+      - `rook/csi/plugin-pod/count/nfs`
+
+### CephCluster metrics
+- `rook/cluster/node/count/with-ceph-pods` - Number of nodes on which the cluster has running pods.
```

Comment:
Shall we merge with the k8s node metrics? 
```suggestion
- `rook/kubernetes/node/count/with-ceph-pods` - Number of nodes on which the cluster has running pods.
```

Reply:
I think that still seems like a cluster metric to me. It will be different from cluster-to-cluster, so I think that makes it a cluster metric.

Reply:
Everything is a cluster metrics, right? If the operator is managing different clusters, all the metrics are still reported to this instance of ceph, and each instance of ceph would phone home separately to report telemetry. So if we're going to have metrics about nodes, seems like they should all be under the same tree, whether it's `rook/cluster` or `rook/kubernetes` it probably doesn't matter.

Reply:
Sure, but for CSI, it seems straightforward to just count the plugin pods under the `csi` heading. It's really only the "nodes with ceph pods" idea that is a little bit abstract/funny.

Reply:
I also disagree that all node information should be under the same heading. Inherently, the total number of Kubernetes nodes is platform information, which I would think should be under a different heading than information about the Rook-Ceph cluster itself.

But I suppose it depends on whose "view" is most important. Since this is Ceph telemetry, I think Ceph's view is the makes the most sense. In that view, I see platform information separately.

But if we take the Kubernetes view, then everything is in comparison the the Kubernetes cluster, and what you say would make sense. But I don't think that "view" is the right one for Ceph telemetry.